thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/dao/TaskDao.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2022 StarTree Inc
+ *
+ * Licensed under the StarTree Community License (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at http://www.startree.ai/legal/startree-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT * WARRANTIES OF ANY KIND,
+ * either express or implied.
+ * See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package ai.startree.thirdeye.datalayer.dao;
+
+import static java.util.Objects.requireNonNull;
+
+import ai.startree.thirdeye.datalayer.DatabaseService;
+import ai.startree.thirdeye.datalayer.DatabaseTransactionService;
+import ai.startree.thirdeye.datalayer.entity.TaskEntity;
+import ai.startree.thirdeye.datalayer.mapper.TaskEntityMapper;
+import ai.startree.thirdeye.spi.datalayer.DaoFilter;
+import ai.startree.thirdeye.spi.datalayer.Predicate;
+import ai.startree.thirdeye.spi.datalayer.dto.TaskDTO;
+import ai.startree.thirdeye.spi.json.ThirdEyeSerialization;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.apache.commons.collections4.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class TaskDao {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TaskDao.class);
+  private static final boolean IS_DEBUG = LOG.isDebugEnabled();
+  private static final ObjectMapper OBJECT_MAPPER = ThirdEyeSerialization.getObjectMapper();
+
+  private final DatabaseService databaseService;
+  private final DatabaseTransactionService transactionService;
+
+  @Inject
+  public TaskDao(final DatabaseService databaseService,
+      final DatabaseTransactionService transactionService) {
+    this.databaseService = databaseService;
+    this.transactionService = transactionService;
+  }
+
+  private TaskEntity toEntity(final TaskDTO dto)
+      throws JsonProcessingException {
+    TaskEntity entity = TaskEntityMapper.INSTANCE.toTaskEntity(dto);
+    final String jsonVal = toJsonString(dto);
+    entity.setJsonVal(jsonVal);
+    final Timestamp currentTime = new java.sql.Timestamp(System.currentTimeMillis());
+    entity.setCreateTime(currentTime);
+    entity.setUpdateTime(currentTime);
+    entity.setVersion(dto.getVersion() == 0 ? 1 : dto.getVersion());
```

Comment:
What's with this weird logic? so 1 is the final version?

let's not make any changes in this PR. but let's have a JIRA for this

Reply:
Yes that's a good point. The logic is `toEntity` gets called from `put` and `update`, in `put` flow the version needs to be set as 1 (first version) and for `update` take the version provided by `dto`. `put` operations should take the responsibility of setting the version as 1 before calling `toEntity`. Tried to mimic the logic from `GenericPojoDao.toGenericJsonEntity`. Will handle this separately with some tests to make sure there is nothing more to it.